elements/pfe-codeblock/src/pfe-codeblock.js

```diff
@@ -0,0 +1,269 @@
+        type: Number,
+        default: 1
+      },
+      codeTheme: {
+        title: "Code Theme",
+        type: String,
+        values: ["dark", "light"],
+        default: "light"
+      }
+    };
+  }
+
+  static get slots() {
+    return {};
+  }
+
+  get codeblock() {
+    return this._codeblock;
+  }
+
+  set codeblock(text) {
+    if (!text) {
+      return;
+    }
+
+    this._codeblock = text;
+    this.renderCodeblock();
+  }
+
+  // Declare the type of this component
+  static get PfeType() {
+    return PFElement.PfeTypes.Content;
+  }
+
+  constructor() {
+    super(PfeCodeblock, { type: PfeCodeblock.PfeType });
+
+    this._codeblock = null;
+    this._codeblockRender = null;
+    this._codeblockRenderOuterPreTag = null;
+    this._codeblockContainer = null;
+    this._readyStateChangeHandler = this._readyStateChangeHandler.bind(this);
+
+    this._observer = new MutationObserver((mutationList, observer) => {
+      if (!this._codeblockContainer.textContent) {
+        this._codeblockRender.innerHTML = "";
+        return;
+      }
+
+      // TODO: when we stop supporting IE11, the need to disconnect and
+      // then reconnect will no longer be needed
+      if (window.ShadyCSS) {
+        observer.disconnect();
+      }
+
+      this.codeblock = this._codeblockContainer.textContent;
+
+      if (window.ShadyCSS) {
+        this._muationObserve();
+      }
+    });
+  }
+
+  //return class for line numbers
+  lineNumberCssClass() {
+    return this.codeLineNumbers ? " line-numbers" : "";
+  }
+
+  //return a valid prism.js language css class
+  codePrismLanguage() {
+    return "language-" + this.codeLanguage;
+  }
+
+  //return a prism.js language lib
+  codePrismLanguageLoad() {
+    return Prism.languages[this.codeLanguage];
+  }
+
+  //return applied classes for pre
+  appliedCssClasss() {
+    return this.codePrismLanguage() + this.lineNumberCssClass();
+  }
+
+  //apply classes to the markup for prism.js
+  setComponentClasses() {
+    this._codeblockRender.setAttribute("class", this.codePrismLanguage());
+    this._codeblockRenderOuterPreTag.setAttribute("class", this.appliedCssClasss());
+    if (this.codeLineNumberStart !== 1) {
+      this._codeblockRenderOuterPreTag.style.counterReset = "linenumber " + (this.codeLineNumberStart - 1);
+    }
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    //create dom elements and attach language styles
+    this._codeblockRenderOuterPreTag = document.createElement("pre");
+    this._codeblockRender = document.createElement("code");
+    this._codeblockRender.setAttribute("codeblock-render", "");
+    this.setComponentClasses();
+    this._codeblockRenderOuterPreTag.appendChild(this._codeblockRender);
+
+    //Add to shadow-root
+    //this.appendChild(this._codeblockRenderOuterPreTag);
+    this.shadowRoot.appendChild(this._codeblockRenderOuterPreTag);
+
+    this.shadowRoot.querySelector("slot").addEventListener("slotchange", () => {
```

Comment:
The slotchange event listener should be removed in the disconnectedCallback.

Reply:
@kylebuch8 should the slot change be removed altogether or moved?

Reply:
It's fine to remove it.